streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java

```diff
@@ -16,17 +16,240 @@
+        outputStream2 = appId + "-output-stream-2";
+        outputStream3 = appId + "-output-stream-3";
+        storeChangelog1 = appId + "-topology-1-store-changelog";
+        storeChangelog2 = appId + "-topology-2-store-changelog";
+        storeChangelog3 = appId + "-topology-3-store-changelog";
+        props = configProps();
+        CLUSTER.createTopic(inputStream1, 2, 1);
+        CLUSTER.createTopic(inputStream2, 2, 1);
+        CLUSTER.createTopic(inputStream3, 2, 1);
+        CLUSTER.createTopic(outputStream1, 2, 1);
+        CLUSTER.createTopic(outputStream2, 2, 1);
+        CLUSTER.createTopic(outputStream3, 2, 1);
+    }
+
+    @After
+    public void shutdown() throws Exception {
+        if (streams != null) {
+            streams.close(Duration.ofSeconds(30));
+        }
+        CLUSTER.deleteTopics(inputStream1, inputStream2, inputStream3, outputStream1, outputStream2, outputStream3);
+    }
+
+    @Test
+    public void shouldProcessSingleNamedTopologyAndPrefixInternalTopics() throws Exception {
+        produceToInputTopics(inputStream1, standardInputData);
+        builder1.stream(inputStream1)
+            .selectKey((k, v) -> k)
+            .groupByKey()
+            .count(Materialized.as(Stores.persistentKeyValueStore("store")))
+            .toStream().to(outputStream1);
+        streams = new KafkaStreamsNamedTopologyWrapper(builder1.buildNamedTopology(props), props, clientSupplier);
+        IntegrationTestUtils.startApplicationAndWaitUntilRunning(singletonList(streams), Duration.ofSeconds(15));
+        final List<KeyValue<String, Long>> results = waitUntilMinKeyValueRecordsReceived(consumerConfig, outputStream1, 3);
+        assertThat(results, equalTo(standardOutputData));
+
+        final Set<String> allTopics = CLUSTER.getAllTopicsInCluster();
+        assertThat(allTopics.contains(appId + "-" + "topology-1" + "-store-changelog"), is(true));
+        assertThat(allTopics.contains(appId + "-" + "topology-1" + "-store-repartition"), is(true));
+    }
+
+    @Test
+    public void shouldProcessMultipleIdenticalNamedTopologiesWithPersistentStateStores() throws Exception {
+        produceToInputTopics(inputStream1, standardInputData);
+        produceToInputTopics(inputStream2, standardInputData);
+        produceToInputTopics(inputStream3, standardInputData);
+
+        builder1.stream(inputStream1).selectKey((k, v) -> k).groupByKey().count(Materialized.as(Stores.persistentKeyValueStore("store"))).toStream().to(outputStream1);
+        builder2.stream(inputStream2).selectKey((k, v) -> k).groupByKey().count(Materialized.as(Stores.persistentKeyValueStore("store"))).toStream().to(outputStream2);
+        builder3.stream(inputStream3).selectKey((k, v) -> k).groupByKey().count(Materialized.as(Stores.persistentKeyValueStore("store"))).toStream().to(outputStream3);
+        streams = new KafkaStreamsNamedTopologyWrapper(buildNamedTopologies(builder1, builder2, builder3), props, clientSupplier);
```

Comment:
Just curious, what if we just pass in `buildNamedTopologies(builder1, builder1, builder1)` here?

Reply:
If you pass in the same builder object it will end up throwing the `IllegalStateException: Tried to set topologyName but the name was already set` since it's using the same underlying InternalTopologyBuilder. Actually the same thing would have happened before named topologies but for `applicationId` instead of `topologyName` , except that we don't do this check in `setApplicationId` 

Personally I would argue that we should also do this check in `setApplicationId`, but it would probably break hundreds of tests which happen to reuse the same builder across different KafkaStreams. Imo the real issue is that the StreamsBuilder class (or technically the InternalStreamsBuilder) has its own InternalTopologyBuilder that it creates during instantiation and then actively updates as the DSL topology is built, even though it should just be a basic builder class that simply specifies the DSL and the InternalTopologyBuilder is only created when calling StreamsBuilder#build.

It seems like some awkwardness in the logical vs physical topology plan: the StreamsBuilder feels like it should only cover the logical DSL topology, and the StreamsBuilder alone, but in reality this leeches into the InternalTopologyBuilder. Whereas the InternalTopologyBuilder feels like it should just be responsible for the physical plan generated by the logical plan in StreamsBuilder#build, yet in fact this class contains pretty much everything -- it's the logical plan, the physical plan, and all the metadata that Streams needs to access.

Anyways that's my long rant on the InternalTopologyBuilder, I think you actually filed a ticket for cleaning up this logical vs physical plan logic so maybe that will cover it 🙂